db/sqlbase.go

```diff
@@ -0,0 +1,109 @@
+package db
+
+import (
+	"context"
+	"os"
+	"sync"
+
+	"database/sql"
+	"github.com/rs/zerolog"
+
+	// this is required for sqlite3 usage
+	"github.com/iotexproject/iotex-core/pkg/lifecycle"
+)
+
+// Store is the interface of KV store.
+type Store interface {
+	lifecycle.StartStopper
+
+	// Get DB instance
+	GetDB() *sql.DB
+
+	// Transact wrap the transaction
+	Transact(txFunc func(*sql.Tx) error) (err error)
+}
+
+// sqlbase is local sqlite3
+type sqlbase struct {
+	mutex      sync.RWMutex
+	db         *sql.DB
+	connectStr string
+	driverName string
+}
+
+// logger is initialized with default settings
+var logger = zerolog.New(os.Stderr).Level(zerolog.InfoLevel).With().Timestamp().Logger()
+
+// NewSQLBase instantiates an sqlite3
+func NewSQLBase(driverName string, connectStr string) Store {
+	return &sqlbase{db: nil, connectStr: connectStr, driverName: driverName}
+}
+
+// Start opens the SQL (creates new file if not existing yet)
+func (s *sqlbase) Start(_ context.Context) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.db != nil {
+		return nil
+	}
+
+	// Use db to perform SQL operations on database
+	db, err := sql.Open(s.driverName, s.connectStr)
+	if err != nil {
+		return err
+	}
+	s.db = db
+	return nil
+}
+
+// Stop closes the SQL
+func (s *sqlbase) Stop(_ context.Context) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.db != nil {
+		err := s.db.Close()
+		s.db = nil
+		return err
+	}
+	return nil
+}
+
+// Stop closes the SQL
+func (s *sqlbase) GetDB() *sql.DB {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	return s.db
+}
+
+// Transact wrap the transaction
+func (s *sqlbase) Transact(txFunc func(*sql.Tx) error) (err error) {
+	s.mutex.Lock()
```

Comment:
I wonder if we should synchronize transact. DB should have supported concurrency

Reply:
Synchronization might be the requirement from the use case, so that it should be done there

Reply:
make sense